Discovery engine PR Request

LegendaryMercury opened the pull request:
10 policy's and some extra file structures for whoever may wish to take over and finish the resource. Its a big resource so there is a lot left to do!

sumedh004 reviewed (changes requested):
Looks like your plan.json for all didnt got committed into the branch so please have a look and update

sumedh004 commented on policies/gcp/discovery_engine/chat_engine/chat_engine_location/policy.rego:
```diff
@@ -0,0 +1,25 @@
+package terraform.gcp.security.discovery_engine.chat_engine.chat_engine_location # Edit here 
```
Delete the template comments. This is to be done for all files

LegendaryMercury replied:
This has been done.

sumedh004 commented on inputs/gcp/discovery_engine/chat_engine/chat_engine_config/nc.tf:
```diff
@@ -0,0 +1,30 @@
+  }
+
+resource "google_discovery_engine_chat_engine" "nc" {
+  engine_id = "chat-engine-id"
```
In the vars.rego file, whatever is declared as resource_value_name (for example, engine_id in the case ), should be prefixed with either c- or nc- in the corresponding Terraform files.

For example, if resource_value_name is set to engine_id, the value in the Terraform file should include c/nc as:
engine_id = "c"  and so on for c.tf
engine_id = "nc" and so on for nc.tf

This is to be done for all tf files

sumedh004 commented on policies/gcp/discovery_engine/chat_engine/chat_engine_location/policy.rego:
```diff
@@ -0,0 +1,25 @@
+      {
+        "condition": "location is mis-configured",
+        "attribute_path": ["location"],
+        "values": ["eu"],
```
You probably can keep aus locations for whitelisting

LegendaryMercury replied:
AU isn't an option, it has to be US EU or Global. U can check on the site, idk why.

sumedh004 replied:
You can remove location policies if AU isnt an option

sumedh004 commented on policies/gcp/discovery_engine/chat_engine/vars.rego:
```diff
@@ -0,0 +1,10 @@
+variables := {
+    "friendly_resource_name": "Chat Engine",
+    "resource_type":  "google_discovery_engine_chat_engine",
+    "resource_value_name" : "engine_id" # eg., "name"
```
Delete the template comments

LegendaryMercury replied:
This has been done.

sumedh004 commented on policies/gcp/discovery_engine/cmek_config/cmek_config_single_region_keys/policy.rego:
```diff
@@ -0,0 +1,25 @@
+      {
+        "condition": "single_region_keys is mis-configured",
+        "attribute_path": ["single_region_keys", 0, "kms_key"],
+        "values": ["projects/735927692082/locations/eu/keyRings/my-ring/cryptoKeys/my-eu-key", "projects/735927692082/locations/europe-west1/keyRings/my-ring/cryptoKeys/my-ew1-key", "projects/735927692082/locations/europe-north1/keyRings/my-ring/cryptoKeys/my-en1-key"],
```
Pattern whitelist policy would be suited for this

LegendaryMercury replied:
I have planned to make that change

LegendaryMercury replied:
The change is made/

sumedh004 commented on policies/gcp/discovery_engine/data_store/data_store_document_processing_config/policy.rego:
Can you explain security relevance of this policy as this doesnt look security related

LegendaryMercury replied:
Yeah sorry just had to check what i put in my notes:
document_processing_config(Optional), high concern, controls how documents are processed and understood by AI. Can result in AIs misunderstanding documents or grab data you didn't think they were able too, like image meta data. 
that.

sumedh004 commented on policies/gcp/discovery_engine/data_store/data_store_kms_key_name/policy.rego:
```diff
@@ -0,0 +1,24 @@
+      {
+        "condition": "KMS Key is mis-configured",
+        "attribute_path": ["kms_key_name"],
+        "values": ["/project/keys/my-safe-key"],
```
Pattern whitelist better suited for this

LegendaryMercury replied:
Ok i have changed it. But i don't think it likes it, the input is sensitive so this was the only method that worked without throwing errors.

sumedh004 commented on inputs/gcp/discovery_engine/recommendation_engine/recommendation_engine_data_store_ids/nc.tf:
```diff
@@ -0,0 +1,27 @@
+  display_name                 = "Example Recommendation Engine"
+  data_store_ids               = [google_discovery_engine_data_store.notcompliant.data_store_id]
+  industry_vertical            = "GENERIC"
+  common_config {
```
Optional attributes like this (common_config block )are to be removed. Only keep the necessary attributes required to create the resource and policy

LegendaryMercury replied:
I think i removed everything unnecessary now.

sumedh004 commented on policies/gcp/discovery_engine/recommendation_engine/recommendation_engine_data_store_ids/policy.rego:
```diff
@@ -0,0 +1,26 @@
+      {
+        "condition": "data_store_ids is mis-configured",
+        "attribute_path": ["data_store_ids"],
+        "values": ["c-recommendation-datastore-id"],
```
This is not security related. The attribute in tf files takes only id of the data store

LegendaryMercury replied:
data_store_ids - (Required) The data stores associated with this engine. For SOLUTION_TYPE_RECOMMENDATION type of engines, they can only associate with at most one data store. Moderate\High concern, if not configured correctly, data leakage could happen. Make sure private data is not selected. 

That was my reasoning in my notes.

sumedh004 replied:
this isnt working correctly

<img width="875" height="117" alt="image" src="https://github.com/user-attachments/assets/540034f2-59ac-4e4e-88a1-4340f702ec3f" />

sumedh004 commented on policies/gcp/discovery_engine/engine_schema/engine_schema_json/policy.rego:
```diff
@@ -0,0 +1,37 @@
+      #And Block
+        "condition": "Json is mis-configured",
+        "attribute_path": ["json_schema"],
+        "values": ["{\"$schema\":\"https://google.com\"}"],
```
What is being checked here for json? Also looks too restrictive

LegendaryMercury replied:
I think i updated that one.

sumedh004 reviewed (changes requested):
Majority of policies are good apart from a couple which I have asked for explanation

Please include plan.json in all of your tf folders as its missing now.

The following is applicable for all of the policies:


In the vars.rego file, whatever is declared as resource_value_name (for example, in the case ), should be prefixed with either c- or nc- in the corresponding Terraform files.

For example, if resource_value_name is set to rule, the value in the Terraform file should include c/nc as:
rule = c and so on for c.tf
rule = nc  and so on for nc.tf

This should be the naming convention

LegendaryMercury committed:
Fixed Mistakes

LegendaryMercury committed:
Made Some Compliancy Changes

LegendaryMercury committed:
Made Some Requested Changes

Made some Changes

LegendaryMercury committed:
Added Plan Files

Added Plan Files

LegendaryMercury commented:
I made some changes to minor stuff, ill change the policy's themselves at a later date.

LegendaryMercury committed:
Pattern_whitelist

Changed 2 policys to pattern whitelist

LegendaryMercury committed:
Changed_engine_schema_json

PatternWhitelist

LegendaryMercury committed:
Updated_recommendation_engine

Removed unneeded stuff.

sumedh004 commented on policies/gcp/discovery_engine/cmek_config/cmek_config_location/policy.rego:
If AU isnt an option for location then remove this policy

sumedh004 commented on policies/gcp/discovery_engine/cmek_config/cmek_config_single_region_keys/policy.rego:
This policy isnt working properly, nc is passed as well

<img width="888" height="153" alt="image" src="https://github.com/user-attachments/assets/812d26d1-8b72-4d28-b095-90b5aaf45001" />

LegendaryMercury replied:
<img width="747" height="383" alt="image" src="https://github.com/user-attachments/assets/66a4dd8e-bb5b-49e1-a891-684892e820c2" />

Should be OK next push, pattern whitelist isnt working.

sumedh004 commented on policies/gcp/discovery_engine/data_store/data_store_document_processing_config/policy.rego:
```diff
@@ -0,0 +1,26 @@
+        "remedies": ["Ensure that it is set to true"]
+        },
+      {
+        "condition": "allow_cross_region is mis-configured",
```
Update the condition message according to the policy.

LegendaryMercury replied:
Fixed

sumedh004 commented on policies/gcp/discovery_engine/data_store/data_store_location/policy.rego:
Remove this policy if AU isnt an option

sumedh004 commented on policies/gcp/discovery_engine/engine_schema/engine_schema_json/policy.rego:
<img width="887" height="97" alt="image" src="https://github.com/user-attachments/assets/5fb5e34a-0cd9-4102-b297-6e86fce0baeb" />

Policy failing check

LegendaryMercury replied:
Dam ill have to re-write that one.

LegendaryMercury replied:
<img width="1229" height="370" alt="image" src="https://github.com/user-attachments/assets/c066adee-4b45-445a-85cb-d7d944bcc9af" />
Is this the same one, its working for me?

sumedh004 commented on policies/gcp/discovery_engine/engine_schema/engine_schema_location/policy.rego:
Remove if AU isnt an option

sumedh004 reviewed (changes requested):
For locations if AU isnt supported option then you can remove those policies.

Also a couple of policies are flagging both c and nc so check them as well.

I still need some clarification for the policy engine_schema_json in whats being checked

Rest is good

LegendaryMercury committed:
Taking out Trash

Gone
(the location policys that arent in AUS)
(also file adjustments for the engine_schema json check, back to whitelist but with full expression)

LegendaryMercury commented:
I have come to agree that recomendation_data_store_ids isn't really much of a security issue. So i also will remove it.

LegendaryMercury committed:
More Changes

More Changes Made

LegendaryMercury committed:
Added location Policy's Back

Yotam Said i could add them, because i did a few they will just be worth less.

LegendaryMercury committed:
misc

some misc changes.

LegendaryMercury committed:
Merge remote-tracking branch 'origin/dev' into gcp/service/discovery_engine

LegendaryMercury committed:
Documentation

Hope this is what you wanted.

LegendaryMercury commented:
I re-added the location policy's. They work fine and the service may come to AU eventually. Till then EU is the best option to host.

LegendaryMercury committed:
Merge remote-tracking branch 'origin/gcp/service/discovery_engine' into dev

sumedh004 commented:
everything is good apart from the industry vertical policy which is failing

<img width="1048" height="131" alt="image" src="https://github.com/user-attachments/assets/78a25e1b-a657-42e1-9b82-fa42edc1a9fd" />

LegendaryMercury committed:
search_engine_fix

LegendaryMercury committed:
engine_config_fix

LegendaryMercury committed:
Opps

Opps

sumedh004 reviewed (dismissed):
LGTM

sumedh004 reviewed (dismissed):
LGTM

Pat-Stuart committed:
Merge branch 'dev' into gcp/service/discovery_engine

sumedh004 reviewed (approved):
LGTM
